File: clients/new-js/packages/chromadb/src/sync-types.ts
import type {
  Source as GeneratedSource,
  Job as GeneratedJob,
  JobStatus as GeneratedJobStatus,
  SourceChunkingConfig as GeneratedChunkingConfig,
  GitRefIdentifier as GeneratedGitRefIdentifier,
  SourceTypeFilter as GeneratedSourceTypeFilter,
  OrderBy as GeneratedOrderBy,
  AutoSync as GeneratedAutoSync,
  SparseEmbeddingModel as GeneratedSparseEmbeddingModel,
  HashMap as GeneratedHashMap,
} from "./sync-api";

// Re-export generated types that are already well-shaped
export type SyncSource = GeneratedSource;
export type Invocation = GeneratedJob;
export type InvocationStatus = GeneratedJobStatus;

/**
 * Only the string variants of InvocationStatus are valid as query parameter
 * filters. Object variants like `{ complete: ... }` cannot be serialized as
 * URL query parameters.
 */
export type InvocationStatusFilter = Extract<InvocationStatus, string>;
export type SyncChunkingConfig = GeneratedChunkingConfig;
export type GitRefIdentifier = GeneratedGitRefIdentifier;
export type SourceTypeFilter = GeneratedSourceTypeFilter;
export type SyncOrderBy = GeneratedOrderBy;
export type AutoSyncMode = GeneratedAutoSync;
export type SparseEmbeddingModel = GeneratedSparseEmbeddingModel;
export type InvocationsByKeysResult = GeneratedHashMap;

export const DenseEmbeddingModel = {
  Qwen3Embedding06B: "Qwen/Qwen3-Embedding-0.6B",
} as const;
export type DenseEmbeddingModelValue =
  (typeof DenseEmbeddingModel)[keyof typeof DenseEmbeddingModel];

export const SparseEmbeddingModel = {
  BM25: "Chroma/BM25",
  SpladeV1: "prithivida/Splade_PP_en_v1",
} as const;
export type SparseEmbeddingModelValue =
  (typeof SparseEmbeddingModel)[keyof typeof SparseEmbeddingModel];

export interface SyncEmbeddingConfig {
  dense?: {
    model: DenseEmbeddingModelValue;
    task?: { taskName: string; queryPrompt?: string; documentPrompt?: string };
  };
  sparse?: {
    model?: SparseEmbeddingModelValue;
    key?: string;
  } | null;
}

export interface GitHubSourceConfig {
  repository: string;
  appId?: string;
  includeGlobs?: string[];
}

export interface S3SourceConfig {
  bucketName: string;
  region: string;
  collectionName: string;
  awsCredentialId: number;
  pathPrefix?: string;
  autoSync?: AutoSyncMode;
}

export interface WebSourceConfig {
  startingUrl: string;
  maxDepth?: number;
  pageLimit?: number;
  includePathRegexes?: string[];
  excludePathRegexes?: string[];
}

export interface CreateSourceBase {
  databaseName: string;
  embedding?: SyncEmbeddingConfig;
  chunking?: SyncChunkingConfig;
}

export interface CreateGitHubSourceArgs extends CreateSourceBase {
  github: GitHubSourceConfig;
}

export interface CreateS3SourceArgs extends CreateSourceBase {
  s3: S3SourceConfig;
}

export interface CreateWebSourceArgs extends CreateSourceBase {
  web: WebSourceConfig;
}

export interface CreateGitHubInvocationArgs {
  targetCollectionName: string;
  refIdentifier: GitRefIdentifier;
}

export interface CreateS3InvocationArgs {
  objectKey: string;
  targetCollectionName?: string;
  customId?: string;
  metadata?: Record<string, unknown>;
}

export interface CreateWebInvocationArgs {
  targetCollectionName: string;
}

export type CreateInvocationArgs =
  | CreateGitHubInvocationArgs
  | CreateS3InvocationArgs
  | CreateWebInvocationArgs;

export interface ListOptions {
  limit?: number;
  offset?: number;
  orderBy?: SyncOrderBy;
}

export interface ListSourcesOptions extends ListOptions {
  databaseName?: string;
  sourceType?: SourceTypeFilter;
}

export interface ListInvocationsOptions extends ListOptions {
  sourceId?: string;
  databaseName?: string;
  sourceType?: SourceTypeFilter;
  status?: InvocationStatusFilter;
}

export interface SyncClientArgs {
  apiKey?: string;
  host?: string;
  fetchOptions?: Omit<RequestInit, "headers"> & {
    headers?: Record<string, string>;
  };
}

File: chromadb/sync_types.py
from __future__ import annotations

import enum
from dataclasses import dataclass
from typing import Any, Dict, List, Literal, Optional, Union


# --- Enums ---


class DenseEmbeddingModel(str, enum.Enum):
    QWEN3_EMBEDDING_06B = "Qwen/Qwen3-Embedding-0.6B"


class SparseEmbeddingModel(str, enum.Enum):
    BM25 = "Chroma/BM25"
    SPLADE_V1 = "prithivida/Splade_PP_en_v1"


class InvocationStatus(str, enum.Enum):
    PENDING = "pending"
    PROCESSING = "processing"
    COMPLETE = "complete"
    FAILED = "failed"
    CANCELLED = "cancelled"


class SourceTypeFilter(str, enum.Enum):
    GITHUB = "github"
    WEB_SCRAPE = "web_scrape"
    S3 = "s3"


class OrderBy(str, enum.Enum):
    ASC = "ASC"
    DESC = "DESC"


class AutoSyncMode(str, enum.Enum):
    NONE = "none"
    DIRECT = "direct"
    METADATA = "metadata"


# --- Embedding config ---


@dataclass
class EmbeddingTask:
    task_name: str
    query_prompt: Optional[str] = None
    document_prompt: Optional[str] = None


@dataclass
class DenseEmbeddingConfig:
    model: DenseEmbeddingModel
    task: Optional[EmbeddingTask] = None


@dataclass
class SparseEmbeddingConfig:
    model: Optional[SparseEmbeddingModel] = None
    key: Optional[str] = None


@dataclass
class SyncEmbeddingConfig:
    dense: Optional[DenseEmbeddingConfig] = None
    sparse: Optional[SparseEmbeddingConfig] = None


# --- Chunking config ---


@dataclass
class TreeSitterChunking:
    type: Literal["tree_sitter"] = "tree_sitter"
    max_size_bytes: Optional[int] = None


@dataclass
class LinesChunking:
    type: Literal["lines"] = "lines"
    max_lines: Optional[int] = None
    max_size_bytes: Optional[int] = None


ChunkingConfig = Union[TreeSitterChunking, LinesChunking]


# --- Source configs ---


@dataclass
class GitHubSourceConfig:
    repository: str
    app_id: Optional[str] = None
    include_globs: Optional[List[str]] = None


@dataclass
class S3SourceConfig:
    bucket_name: str
    region: str
    collection_name: str
    aws_credential_id: int
    path_prefix: Optional[str] = None
    auto_sync: Optional[AutoSyncMode] = None


@dataclass
class WebSourceConfig:
    starting_url: str
    max_depth: Optional[int] = None
    page_limit: Optional[int] = None
    include_path_regexes: Optional[List[str]] = None
    exclude_path_regexes: Optional[List[str]] = None


# --- Create source args ---


@dataclass
class CreateGitHubSourceArgs:
    github: GitHubSourceConfig
    embedding: Optional[SyncEmbeddingConfig] = None
    chunking: Optional[ChunkingConfig] = None


@dataclass
class CreateS3SourceArgs:
    s3: S3SourceConfig
    embedding: Optional[SyncEmbeddingConfig] = None
    chunking: Optional[ChunkingConfig] = None


@dataclass
class CreateWebSourceArgs:
    web: WebSourceConfig
    embedding: Optional[SyncEmbeddingConfig] = None
    chunking: Optional[ChunkingConfig] = None


# --- Git ref identifier ---

GitRefIdentifier = Union[Dict[Literal["sha"], str], Dict[Literal["branch"], str]]


# --- Create invocation args ---


@dataclass
class CreateGitHubInvocationArgs:
    target_collection_name: str
    ref_identifier: GitRefIdentifier


@dataclass
class CreateS3InvocationArgs:
    object_key: str
    target_collection_name: Optional[str] = None
    custom_id: Optional[str] = None
    metadata: Optional[Dict[str, Any]] = None


@dataclass
class CreateWebInvocationArgs:
    target_collection_name: str


CreateInvocationArgs = Union[
    CreateGitHubInvocationArgs,
    CreateS3InvocationArgs,
    CreateWebInvocationArgs,
]


# --- List options ---


@dataclass
class ListSourcesOptions:
    source_type: Optional[SourceTypeFilter] = None
    limit: Optional[int] = None
    offset: Optional[int] = None
    order_by: Optional[OrderBy] = None


@dataclass
class ListInvocationsOptions:
    source_id: Optional[str] = None
    source_type: Optional[SourceTypeFilter] = None
    status: Optional[InvocationStatus] = None
    limit: Optional[int] = None
    offset: Optional[int] = None
    order_by: Optional[OrderBy] = None


# --- Response types ---
# These are returned as dicts from the API. We define type aliases for clarity.

SyncSource = Dict[str, Any]
Invocation = Dict[str, Any]
InvocationsByKeysResult = Dict[str, Any]
